buildenv/buildenv.nix

```diff
@@ -0,0 +1,278 @@
+{
+  activationScripts_storePath ? "@activationScripts@",
+  defaultEnvrc_storePath ? "@defaultEnvrc@",
+  coreutils_storePath ? "@coreutils@",
+  floxBuildenv_storePath ? "@out@",
+  manifest,
+  name ? "environment",
+  serviceConfigYaml ? "",
+}:
+let
+
+  # Start by converting (text) storePaths into contextual strings.
+  # FTR a contextual string is one that knows it's a derivation!
+  # This is extremely important for populating the Nix dependency
+  # graph. See `builtins.getContext` and `addContextFrom` in nixpkgs
+  # strings.nix.
+  activationScripts = builtins.storePath activationScripts_storePath;
+  defaultEnvrc = builtins.storePath defaultEnvrc_storePath;
+  floxBuildEnv = builtins.storePath floxBuildenv_storePath;
+  coreutils = builtins.storePath coreutils_storePath;
+
+  # A helpful library function copied from nixpkgs/lib/attrsets.nix.
+  # foldlAttrs = f: init: set:
+  #   builtins.foldl'
+  #     (acc: name: f acc name set.${name})
+  #     init
+  #     (builtins.attrNames set);
+
+  # The system we're building for.
+  system = builtins.currentSystem;
+
+  # Copy manifest file into the store for access within derivations.
+  manifestLock = /. + manifest;
+
+  # Parse the manifest file.
+  manifestLockData = builtins.fromJSON (builtins.readFile manifest);
+  manifestData = manifestLockData.manifest;
+
+  build = if (builtins.hasAttr "build" manifestData) then manifestData.build else { };
+  hook = if (builtins.hasAttr "hook" manifestData) then manifestData.hook else { };
+  profile = if (builtins.hasAttr "profile" manifestData) then manifestData.profile else { };
+  vars =
+    if (builtins.hasAttr "vars" manifestData) then
+      (builtins.toFile "envrc-vars" (
+        builtins.concatStringsSep "" (
+          builtins.map (n: "export ${n}=\"${builtins.getAttr n manifestData.vars}\"\n") (
+            builtins.attrNames manifestData.vars
+          )
+        )
+        # alternative ... worth it?
+        #      foldlAttrs (
+        #        acc: n: v: acc + "export ${n}=\"${v}\"\n"
+        #      ) "" manifestData.vars
+      ))
+    else
+      null;
+
+  # Calculate environment outputs.
+  environmentOutputs = [
+    "runtime"
+    "develop"
+  ] ++ (builtins.map (x: "build-${x}") (builtins.attrNames build));
+
+  createManifestChunks =
```

Comment:
suggestion(nb): use a (simplified) version of [`linkFarm`](https://github.com/hsjobeki/nixpkgs/blob/migrate-doc-comments/pkgs/build-support/trivial-builders/default.nix#L499-L554) for this.
In my experience these are easier to follow as to which files are being linked where compared to doing that manually.
Functionally they are similar to waht we have with the distinction that `linkFarm` uses links rather than copying files.

Reply:
I actually have the opposite view which is a desire to reduce the number of paths involved in an environment, for performance reasons and for the clarity of understanding that an environment is composed of its a) interpreter, b) manifest (personalization) package, and c) any packages included in the `[install]` block of the manifest.

It would be different if any of the files included in the manifest package were useful on their own, but as it happens every file in the manifest package is derived from the [1] `manifest.lock` also contained in that package, so the entire set of files must be internally consistent by definition, and therefore not useful outside of that one context.

Reply:
> how are copies of files more performant than symlinks (at a meaningful scale)?

It boils down to the time and resources required to create extra entries in the Nix database, and extra inodes (directories, links) that might not otherwise be required. Having run enormous Nix stores that have run out of inodes in particular I'm always looking for ways to cut the inode count, and if that makes everything else that much smaller and (perhaps imperceptibly) faster I'll take it!

Thanks, I'll go ahead and resolve the conversation.